cpp/src/lists/contains.cu

```diff
@@ -32,308 +30,256 @@
+  __device__ size_type operator()(list_device_view list, thrust::optional<Element> key_opt) const
+  {
+    // A null list never contains any key, even null key.
+    // In addition, a null list will result in a null output row.
+    if (list.is_null()) { return NULL_SENTINEL; }
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
-  }
+    // A null key will also result in a null output row.
+    if (!key_opt) { return NULL_SENTINEL; }
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+    auto const& key = key_opt.value();
+
+    return find_option == duplicate_find_option::FIND_FIRST
+             ? search_list<true, Element>(list, key)
+             : search_list<false, Element>(list, key);
   }
-}
 
-/**
- * @brief __device__ functor to search for a key in a `list_device_view`.
- */
-template <duplicate_find_option find_option>
-struct finder {
-  template <typename ElementType>
-  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+  template <typename Element, CUDF_ENABLE_IF(cudf::is_nested<Element>())>
+  __device__ size_type operator()(list_device_view, thrust::optional<Element>) const
   {
-    auto const list_begin = find_begin<ElementType, find_option>(list);
-    auto const list_end   = find_end<ElementType, find_option>(list);
-    auto const find_iter  = thrust::find_if(
-      thrust::seq, list_begin, list_end, [search_key] __device__(auto element_and_validity) {
-        auto [element, element_is_valid] = element_and_validity;
-        return element_is_valid && cudf::equality_compare(element, search_key);
-      });
-    return distance<find_option>(list_begin, list_end, find_iter);
-  };
-};
+    cudf_assert(false && "Nested types are not yet supported");
+  }
 
-/**
- * @brief Functor to search each list row for the specified search keys.
- */
-template <bool search_keys_have_nulls>
-struct lookup_functor {
-  template <typename ElementType>
-  struct is_supported {
-    static constexpr bool value =
-      cudf::is_numeric<ElementType>() || cudf::is_chrono<ElementType>() ||
-      cudf::is_fixed_point<ElementType>() || std::is_same_v<ElementType, cudf::string_view>;
-  };
-
-  template <typename ElementType, typename... Args>
-  std::enable_if_t<!is_supported<ElementType>::value, std::unique_ptr<column>> operator()(
-    Args&&...) const
+ private:
+  template <bool forward, typename Element, CUDF_ENABLE_IF(is_supported_non_nested_type<Element>())>
+  static __device__ size_type search_list(list_device_view const& list, Element const& search_key)
```

Comment:
I strongly recommend against passing thread local objects by reference in device code as it can impact local memory usage if the compiler decides that an ODR-use requires it to put it in memory whose address can be taken. 
```suggestion
  static __device__ size_type search_list(list_device_view const list, Element const search_key)
```

Reply:
I was once recommended to pass user-defined objects by ref so the amount of mem copy can be minimized.